doc/design/design_high-available_sqlflow_server.md

```diff
@@ -0,0 +1,231 @@
+# High-available SQLFlow Server
+
+## Motivations
+
+In the current system, the SQLFlow client connects the SQLFlow server with a long live connection.
+the SQLFlow client sends a gRPC request which contains a SQL statement and wait until the SQLFlow server finishes executing the SQL statement.
+
+When the SQLFlow server receives one training SQL statement, it will generate a python training program that runs on the host or submit the training job to some distributed training service cluster (Kubernetes/Yarn). This will cause:
+
+1. The local job may cause the SQLFlow server resource insufficient when there are too many SQL jobs.
+1. Sometimes, the SQL job takes too much time, and the gRPC calls timeout.
+1. If one of the SQLFlow server instance fails, the SQL job also fails.
+
+In this design, we propose to:
+
+1. Communicate with the SQLFlow server in a polling manner.
+1. Implement `ClusterJobRunner` on the server-side to launch the SQL Job on Kubernetes.
+
+We recommend using `ClusterJobRunner` in the production environment.
+
+## High-Level Design
+
+The high-availabe SQLFlow job workflow is as follows:
+
+<img src="figures/cluster_job_runner.png">
+
+1. SQLFlow client sends the SQL statement via a gRPC call to the SQLFlow server.
+1. For the `LocalJobRunner`:
+    1. SQLFLow server launches a SQL job on the host and generates a token identifies the SQL job.
+    1. SQLFLow server maintains a mapping from token to the SQL job.
+    1. SQLFlow server returns the token to the client.
+1. For the `ClusterJobRunner`:
+    1. SQLFlow server launches a Kubernetes Pod via Kubernetes API and executes the SQL job in it.
+    1. SQLFlow server fetches the Pod ID.
+    1. SQLFlow server returns the Pod ID as the token to the client.
+1. The SQLFlow client fetches the job result and job status in a polling manner until the returned job status is **SUCCESS** or **FAILED**.
+
+## Proposal Details
+
+### SQLFLow Client
+
+The client calls `Run` and receives a string token. The client subsequently fetches the result using the token periodically. And the client is unaware of the deployment type of the server.
+
+```python
+token = client.Run("sqls ...")
```

Comment:
This doesn't look like a token, but a job ID.

Reply:
You are right. @Yancey1989 shall we change it to `job_id`?